File: mbfpp/code/ScalaExamples/src/main/scala/com/mblinn/mbfpp/functional/fb/DiscountBuilder.scala
package com.mblinn.mbfpp.functional.fb

object DiscountBuilder {

  def discount(percent: Double) = {
    if(percent < 0.0 || percent > 100.0) 
      throw new IllegalArgumentException("Discounts must be between 0.0 and 100.0.")
    (originalPrice: Double) => 
      originalPrice - (originalPrice * percent * 0.01)
  }
  
}

File: mbfpp/code/ClojureExamples/test/mbfpp/test/functional/fb/discount_builder.clj
(ns mbfpp.test.functional.fb.discount-builder
  (:use [mbfpp.functional.fb.discount-builder])
  (:use [midje.sweet]))

(fact "Discount creates functions that calculate 0 percent discounts"
      ((discount 0) 100) => 100.0)

(fact "Discount creates functions that calcuate 100 percent discounts"
      ((discount 100) 100) => 0.0)

(fact "Discount calculates discounts between 1 and 99 percent"
      (doseq [percentage (range 1 100)]
        ((discount percentage) 100) => (- 100 (* percentage 100 0.01))))

(fact "Discounts can only be between 0 and 100"
      (discount -1) => (throws AssertionError)
      (discount 101) => (throws AssertionError))